src/librustc_trans/type_of.rs

```diff
@@ -9,231 +9,472 @@
-    name
+    fn llvm_field_index(&self, index: usize) -> u64 {
+        match self.abi {
+            layout::Abi::Scalar(_) |
+            layout::Abi::ScalarPair(..) => {
+                bug!("TyLayout::llvm_field_index({:?}): not applicable", self)
+            }
+            _ => {}
+        }
+        match self.fields {
+            layout::FieldPlacement::Union(_) => {
+                bug!("TyLayout::llvm_field_index({:?}): not applicable", self)
+            }
+
+            layout::FieldPlacement::Array { .. } => {
+                index as u64
+            }
+
+            layout::FieldPlacement::Arbitrary { .. } => {
+                1 + (self.fields.memory_index(index) as u64) * 2
+            }
+        }
+    }
+
+    fn llvm_pointee<'a>(&self, ccx: &CrateContext<'a, 'tcx>, offset: Size) -> Option<LlvmPointee> {
+        if let Some(&pointee) = ccx.llpointees().borrow().get(&(self.ty, offset)) {
+            return pointee;
+        }
+
+        let mut result = None;
+        match self.ty.sty {
+            ty::TyRawPtr(mt) if offset.bytes() == 0 => {
+                let (size, align) = ccx.size_and_align_of(mt.ty);
+                result = Some(LlvmPointee {
+                    size,
+                    align,
+                    safe: None
+                });
+            }
+
+            ty::TyRef(_, mt) if offset.bytes() == 0 => {
+                let (size, align) = ccx.size_and_align_of(mt.ty);
+
+                let kind = match mt.mutbl {
+                    hir::MutImmutable => if ccx.shared().type_is_freeze(mt.ty) {
+                        PointerKind::Frozen
+                    } else {
+                        PointerKind::Shared
+                    },
+                    hir::MutMutable => {
+                        if ccx.shared().tcx().sess.opts.debugging_opts.mutable_noalias ||
+                           ccx.shared().tcx().sess.panic_strategy() == PanicStrategy::Abort {
+                            PointerKind::UniqueBorrowed
+                        } else {
+                            PointerKind::Shared
+                        }
+                    }
+                };
+
+                result = Some(LlvmPointee {
+                    size,
+                    align,
+                    safe: Some(kind)
+                });
+            }
+
+            _ => {
+                let mut data_variant = match self.variants {
+                    layout::Variants::NicheFilling { dataful_variant, .. } => {
+                        Some(self.for_variant(ccx, dataful_variant))
+                    }
+                    _ => Some(*self)
+                };
+
+                if let Some(variant) = data_variant {
+                    // We're not interested in any unions.
+                    if let layout::FieldPlacement::Union(_) = variant.fields {
+                        data_variant = None;
+                    }
+                }
+
+                if let Some(variant) = data_variant {
```

Comment:
I need add a check here to make sure we only support this for the niche, yeah, good catch! (right now this isn't used anywhere it could create a problem, but it might in the future)